### ISSUE TITLE
mpi_fortran_argv_null_ has incorrect st_size in ompi/mpi/fortran/base/gen-mpi-mangling.pl

### ISSUE THREAD

**Issue by MaskRay** [open-mpi/ompi#7209]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
HEAD (commit ff480706cace1743a8fd7e67d41df6af9a66503b). This bug is also present in openmpi 3.1.0 (built from source), Debian testing  libopenmpi-dev 3.1.3-11 and Arch Linux extra/openmpi 4.0.2-2

## Details of the problem

The C source definition of mpi_fortran_argvs_null_ (` ompi/include/mpif-c-constants.h`) has a st_size field of 8.
```
#  ompi/.libs/libompi_mpir.a(libompi_mpir_la-ompi_mpi_init.o)
% readelf -Ws ompi/.libs/libompi_mpir.a G mpi_fortran_argvs_null_  | grep mpi_fortran_argvs_null_
   212: 0000000000000008     8 OBJECT  GLOBAL DEFAULT  COM mpi_fortran_argvs_null_
#  ompi/.libs/libompi_mpir.a(libompi_mpir_la-ompi_mpi_init.o) is used to link libmpi.so
% readelf -W --dyn-syms ompi/.libs/libmpi.so | grep mpi_fortran_argvs_null_
   666: 000000000012d130     8 OBJECT  GLOBAL DEFAULT   24 mpi_fortran_argvs_null_
```

The Fortran source of mpi_fortran_argvs_null_ has (`ompi/include/mpif-f08-types.h`) a st_size field of 1.
```
% readelf -Ws ompi/mpi/fortran/use-mpi-f08/.libs/cart_create_f08.o | grep mpi_fortran_argvs_null_
    12: 0000000000000010     1 OBJECT  GLOBAL DEFAULT  COM mpi_fortran_argvs_null_

# cart_create_f08.o is used to link libmpi_usempif08.so

# a) If the linker is lld
% readelf -W --dyn-syms ompi/mpi/fortran/use-mpi-f08/.libs/libmpi_usempif08.so | grep mpi_fortran_argvs_null_ 
   639: 0000000000040e30     1 OBJECT  GLOBAL DEFAULT   24 mpi_fortran_argvs_null_

# b) If the linker is GNU ld
% readelf -W --dyn-syms ompi/mpi/fortran/use-mpi-f08/.libs/libmpi_usempif08.so | grep mpi_fortran_argvs_null_ 
   957: 000000000003dd00     8 OBJECT  GLOBAL DEFAULT   24 mpi_fortran_argvs_null_
```

The conflict of st_size can cause mpi_fortran_argvs_null_ to lose symbol version information when the linker is GNU ld https://sourceware.org/bugzilla/show_bug.cgi?id=25236 . For example, Debian shipped packages (libopenmpi-dev) have symbol versions and such libraries cannot be linked by lld due to its rigorous check https://bugs.llvm.org/show_bug.cgi?id=43748

```
% touch a.f90
% mpifort -fuse-ld=lld a.f90
ld.lld: error: corrupt input file: version definition index 0 for symbol mpi_fortran_argv_null_ is out of bounds
>>> defined in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempif08.so
...
```

The correct fix is to make sh_size 8 for the Fortran source.

All symbols that have the problem: `mpi_fortran_argv_null_, mpi_fortran_weights_empty_, mpi_fortran_unweighted_, mpi_fortran_errcodes_ignore_, mpi_fortran_argvs_null_`

**Comment by jsquyres** [comment]
@MaskRay We have arrived at a good solution on PR #7210.  I'll back port this to the v4.0.x release branch.  Do you need it fixed in v3.0.x and v3.1.x, as well?  We do have upcoming releases pending for these series, but we're trying to raise the bar for what constitutes a "really should be fixed" bug on these older series -- i.e., we're trying to do less work on v3.0.x and v3.1.x because they're getting fairly old at this point.

If you need them, I'm happy to back-port this fix to those branches.  I just wanted to ask the question before doing that.

Thanks!

**Comment by MaskRay** [comment]
Thanks. I don't need it to be back ported to v3.0.x and v3.1.x.

@mwglass Do you need it? I assume you filed https://bugs.llvm.org/show_bug.cgi?id=43748

**Comment by mwglass** [comment]
I don’t need it back ported - thanks!

**Comment by mwglass** [comment]
Will this make it into 4.0.3?

**Comment by jsquyres** [comment]
@mwglass Yes, this was merged into the v4.0.x branch.  It will be in the upcoming v4.0.3 release.  You can give it a whirl with a nightly snapshot tarball: https://www.open-mpi.org/nightly/v4.0.x/

Thank you for brining this to our attention!
